File: es6/es6.py
#!/usr/bin/env python3

import sys

def distance(x0, y0, x1, y1):
    d = 0
    if x1 > x0:
        d += x1 - x0
    else:
        d += x0 - x1

    if y1 > y0:
        d += y1 - y0
    else:
        d += y0 - y1
    return d

class Location:
    def __init__(self, id, x, y):
        self.id = id
        self.x = x
        self.y = y

class Region:
    def __init__(self, locs):
        xmin = 1000000
        ymin = 1000000
        xmax = -1000000
        ymax = -1000000
        for l in locs:
            if xmin > l.x:
                xmin = l.x
            if xmax < l.x:
                xmax = l.x
            if ymin > l.y:
                ymin = l.y
            if ymax < l.y:
                ymax = l.y
        self.x = xmin
        self.y = ymin
        self.w = xmax - xmin + 1
        self.h = ymax - ymin + 1
        self.m = [[0 for x in range(self.w)] for y in range(self.h)]

    def partition(self, locs):
        for y in range(self.h):
            for x in range(self.w):
                label = -1
                mind = 10000
                for l in locs:
                    d = distance(x+self.x, y+self.y, l.x, l.y)
                    if d == mind:
                        label = -1
                    elif d < mind:
                        label = l.id
                        mind = d
                self.m[y][x] = label

    def find_area(self, loc):
        area = 0
        for y in range(self.h):
            for x in range(self.w):
                if self.m[y][x] == loc.id:
                    if x == 0 or x == self.w-1 or y == 0 or y == self.h-1:
                        return 0
                    area += 1

        return area

    def find_largest_area(self, locs):
        self.partition(locs)
        largest = 0
        for l in locs:
            area = self.find_area(l)
            if area > largest:
                largest = area
        return largest

    def compute_distance_sums(self, locs):
        for y in range(self.h):
            for x in range(self.w):
                dsum = 0
                for l in locs:
                    dsum += distance(x+self.x, y+self.y, l.x, l.y)
                self.m[y][x] = dsum

    def find_area_lt(self, limit):
        area = 0
        for y in range(self.h):
            for x in range(self.w):
                if self.m[y][x] < limit:
                    area += 1
        return area

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} filename".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    locs = []
    try:
        with open(sys.argv[1], "rt") as f:
            for line in f:
                arr = line.split(",")
                id = len(locs)
                x = int(arr[0])
                y = int(arr[1])
                locs.append(Location(id, x, y))
    except Exception as e:
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    r = Region(locs)
    print("Largest area that isn't infinite:", r.find_largest_area(locs))

    r.compute_distance_sums(locs)
    print("Size of the region with distance sum < 32:", r.find_area_lt(32))
    print("Size of the region with distance sum < 10000:", r.find_area_lt(10000))

File: es6/es6.go
package main

import (
	"fmt"
	"os"
)

const MaxInt = int(^uint(0) >> 1)
const MinInt = -MaxInt - 1

type Location struct {
	Id int
	X  int
	Y  int
}

type Region struct {
	X int
	Y int
	W int
	H int

	Map [][]int
}

func distance(x0, y0, x1, y1 int) int {
	d := 0
	if x1 > x0 {
		d += x1 - x0
	} else {
		d += x0 - x1
	}

	if y1 > y0 {
		d += y1 - y0
	} else {
		d += y0 - y1
	}
	return d
}

func (r *Region) partition(locs []Location) {
	for y := 0; y < r.H; y++ {
		for x := 0; x < r.W; x++ {
			mind := MaxInt
			label := -1
			for _, l := range locs {
				d := distance(x+r.X, y+r.Y, l.X, l.Y)
				if d == mind {
					label = -1
				} else if d < mind {
					mind = d
					label = l.Id
				}
			}
			r.Map[y][x] = label
		}
	}
}

func (r *Region) FindArea(loc *Location) int {
	area := 0
	for y := 0; y < r.H; y++ {
		for x := 0; x < r.W; x++ {
			if loc.Id == r.Map[y][x] {
				if x == 0 || x == r.W-1 || y == 0 || y == r.H-1 {
					return 0
				}
				area++
			}
		}
	}
	return area
}

func (r *Region) FindLargestArea(locs []Location) int {
	r.partition(locs)
	maxarea := MinInt
	for _, l := range locs {
		a := r.FindArea(&l)
		if maxarea < a {
			maxarea = a
		}
	}
	return maxarea
}

func (r *Region) ComputeDistanceSums(locs []Location) {
	for y := 0; y < r.H; y++ {
		for x := 0; x < r.W; x++ {
			sum := 0
			for _, l := range locs {
				sum += distance(x+r.X, y+r.Y, l.X, l.Y)
			}
			r.Map[y][x] = sum
		}
	}
}

func (r *Region) FindAreaLt(limit int) int {
	area := 0
	for y := 0; y < r.H; y++ {
		for x := 0; x < r.W; x++ {
			if r.Map[y][x] < limit {
				area++
			}
		}
	}
	return area
}

func MakeRegion(locs []Location) Region {
	xmin := MaxInt
	ymin := MaxInt
	xmax := MinInt
	ymax := MinInt
	for _, l := range locs {
		if xmin > l.X {
			xmin = l.X
		}
		if xmax < l.X {
			xmax = l.X
		}
		if ymin > l.Y {
			ymin = l.Y
		}
		if ymax < l.Y {
			ymax = l.Y
		}
	}
	w := xmax - xmin + 1
	h := ymax - ymin + 1
	m := make([][]int, h)
	for i := 0; i < h; i++ {
		m[i] = make([]int, w)
	}
	return Region{xmin, ymin, w, h, m}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s filename\n", os.Args[0])
		os.Exit(-1)
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		os.Exit(-1)
	}

	locs := []Location{}
	var loc Location
	for {
		_, err := fmt.Fscanf(input, "%d, %d\n", &loc.X, &loc.Y)
		if err != nil {
			break
		}
		loc.Id = len(locs)
		locs = append(locs, loc)
	}
	input.Close()

	r := MakeRegion(locs)
	fmt.Printf("Largest area that isn't infinite: %d\n", r.FindLargestArea(locs))

	r.ComputeDistanceSums(locs)
	fmt.Printf("Size of the region with distance sum < 32: %d\n",
		r.FindAreaLt(32))
	fmt.Printf("Size of the region with distance sum < 10000: %d\n",
		r.FindAreaLt(10000))

}
